Node keys are ignored when the field is a map

Given this DataFrame 
```
root
 |-- <rel.id>: long (nullable = false)
 |-- <rel.type>: string (nullable = false)
 |-- <source>: map (nullable = false)
 |    |-- key: string
 |    |-- value: string (valueContainsNull = true)
 |-- <target>: map (nullable = false)
 |    |-- key: string
 |    |-- value: string (valueContainsNull = true)
 |-- rel.roles: array (nullable = true)
 |    |-- element: string (containsNull = true)
```


```
+--------+----------+-------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------+
|<rel.id>|<rel.type>|<source>                                                                       |<target>                                                                                                                                                                                                                                                                    |rel.roles                          |
+--------+----------+-------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------+
|7       |ACTED_IN  |[name -> "Emil Eifrem", born -> 1978, <labels> -> ["Person"], <id> -> 8]       |[title -> "The Matrix", released -> 1999, <labels> -> ["Movie"], <id> -> 0, tagline -> "Welcome to the Real World"]                                                                                                                                                         |[Emil]                             |
|3       |ACTED_IN  |[name -> "Hugo Weaving", born -> 1960, <labels> -> ["Person"], <id> -> 4]      |[title -> "The Matrix", released -> 1999, <labels> -> ["Movie"], <id> -> 0, tagline -> "Welcome to the Real World"]                                                                                                                                                         |[Agent Smith]                      |
```

When I try to write the dataframe:

```
import org.apache.spark.sql.{SaveMode, SparkSession}


df2.write
  .format("org.neo4j.spark.DataSource")
  .mode(SaveMode.ErrorIfExists)
  .option("url", "bolt://neo4j:7687")
  .option("database","prova3")
   .option("authentication.type", "basic")
    .option("authentication.basic.username", "neo4j")
    .option("authentication.basic.password", "zeppelin")
  .option("relationship", "ACTED_IN")
    .option("relationship.source.labels", ":Person")
    .option("relationship.source.save.mode", "Overwrite")
    .option("relationship.source.node.keys", "`<source>`.`name`:name")
    .option("relationship.target.labels", ":Movie")
    .option("relationship.target.save.mode", "Overwrite")
    .option("relationship.target.node.keys", "`<target>`.`title`:title")
    .save()
```

I get the following error:

```
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 91.0 failed 1 times, most recent failure: Lost task 0.0 in stage 91.0 (TID 91, localhost, executor driver): java.lang.IllegalArgumentException: Write failed due to the following errors:
 - Schema is missing `<target>`.`title` from option `relationship.target.node.keys`
 - Schema is missing `<source>`.`name` from option `relationship.source.node.keys`

The option key and value might be inverted.
```

--- a/common/src/main/scala/org/neo4j/spark/reader/BasePartitionReader.scala
+++ b/common/src/main/scala/org/neo4j/spark/reader/BasePartitionReader.scala
@@ -24,7 +24,7 @@ abstract class BasePartitionReader(private val options: Neo4jOptions,
   private val driverCache: DriverCache = new DriverCache(options.connection,
     if (partitionSkipLimit.partitionNumber > 0) s"$jobId-${partitionSkipLimit.partitionNumber}" else jobId)
 
-  private val query: String = new Neo4jQueryService(options, new Neo4jQueryReadStrategy(filters, partitionSkipLimit, requiredColumns.getFieldsName))
+  private val query: String = new Neo4jQueryService(options, new Neo4jQueryReadStrategy(filters, partitionSkipLimit, requiredColumns.fieldNames))
     .createQuery()
 
   private val mappingService = new MappingService(new Neo4jReadMappingStrategy(options, requiredColumns), options)

--- a/common/src/main/scala/org/neo4j/spark/service/MappingService.scala
+++ b/common/src/main/scala/org/neo4j/spark/service/MappingService.scala
@@ -30,9 +30,9 @@ class Neo4jWriteMappingStrategy(private val options: Neo4jOptions)
     query(row, schema)
       .forEach(new BiConsumer[String, AnyRef] {
         override def accept(key: String, value: AnyRef): Unit = if (options.nodeMetadata.nodeKeys.contains(key)) {
-          keys.put(key, value)
+          keys.put(options.nodeMetadata.nodeKeys.getOrElse(key, key), value)
         } else {
-          properties.put(key, value)
+          properties.put(options.nodeMetadata.nodeProps.getOrElse(key, key), value)
         }
       })
 

--- a/common/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala
+++ b/common/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala
@@ -1,7 +1,7 @@
 package org.neo4j.spark.util
 
 import javax.lang.model.SourceVersion
-import org.apache.spark.sql.types.{DataTypes, StructField, StructType}
+import org.apache.spark.sql.types.{DataTypes, MapType, StructField, StructType}
 import org.neo4j.driver.types.{Entity, Node, Relationship}
 import org.neo4j.spark.service.SchemaService
 import org.apache.spark.sql.sources.{EqualNullSafe, EqualTo, Filter, GreaterThan, GreaterThanOrEqual, In, IsNotNull, IsNull, LessThan, LessThanOrEqual, Not, StringContains, StringEndsWith, StringStartsWith}
@@ -13,9 +13,11 @@ object Neo4jImplicits {
   implicit class CypherImplicits(str: String) {
     private def isValidCypherIdentifier() = SourceVersion.isIdentifier(str) && !str.trim.startsWith("$")
 
-    def quote(): String = if (!isValidCypherIdentifier() && !str.trim.startsWith("`") && !str.trim.endsWith("`")) s"`$str`" else str
+    def quote(): String = if (!isValidCypherIdentifier() && !str.isQuoted()) s"`$str`" else str
 
     def unquote(): String = str.replaceAll("`", "");
+
+    def isQuoted(): Boolean = str.startsWith("`");
 
     def removeAlias(): String = {
       val splatString = str.split('.')
@@ -98,16 +100,25 @@ object Neo4jImplicits {
   }
 
   implicit class StructTypeImplicit(structType: StructType) {
-    def getFieldsName: Seq[String] = if (structType == null) {
-      Seq.empty
-    } else {
-      structType.map(structField => structField.name)
+    private def isValidMapOrStructField(field: String, structFieldName: String) = {
+      val value: String = """(`.*`)|([^\.]*)""".r.findFirstIn(field).getOrElse("")
+      structFieldName == value.unquote() || structFieldName == value
     }
 
-    def getMissingFields(fields: Set[String]): Set[String] = {
-      val structFieldsNames = structType.getFieldsName
-      fields.filterNot(structFieldsNames.contains(_))
-    }
+    def getMissingFields(fields: Set[String]): Set[String] = fields
+      .map(field => {
+        val maybeField = structType
+          .find(structField => {
+            structField.dataType match {
+              case _: MapType => isValidMapOrStructField(field, structField.name)
+              case _: StructType => isValidMapOrStructField(field, structField.name)
+              case _ => structField.name == field.unquote() || structField.name == field
+            }
+          })
+        field -> maybeField.isDefined
+      })
+      .filterNot(e => e._2)
+      .map(e => e._1)
   }
 
 }

--- a/common/src/main/scala/org/neo4j/spark/util/Neo4jUtil.scala
+++ b/common/src/main/scala/org/neo4j/spark/util/Neo4jUtil.scala
@@ -20,6 +20,7 @@ import org.neo4j.spark.service.SchemaService
 import org.neo4j.spark.util.Neo4jImplicits.EntityImplicits
 import org.slf4j.Logger
 
+import org.neo4j.spark.util.Neo4jImplicits._
 import scala.collection.JavaConverters._
 
 object Neo4jUtil {
@@ -213,7 +214,7 @@ object Neo4jUtil {
 
   def flattenMap(map: java.util.Map[String, AnyRef],
                  prefix: String = ""): java.util.Map[String, AnyRef] = map.asScala.flatMap(t => {
-    val key = if (prefix != "") s"$prefix.${t._1}" else t._1
+    val key: String = if (prefix != "") s"${prefix.quote()}.${t._1.quote()}" else t._1.quote()
     t._2 match {
       case nestedMap: Map[String, AnyRef] => flattenMap(nestedMap.asJava, key).asScala.toSeq
       case _ => Seq((key, t._2))

--- a/common/src/main/scala/org/neo4j/spark/writer/BaseDataWriter.scala
+++ b/common/src/main/scala/org/neo4j/spark/writer/BaseDataWriter.scala
@@ -4,7 +4,7 @@ import org.apache.spark.internal.Logging
 import org.apache.spark.sql.SaveMode
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.types.StructType
-import org.neo4j.driver.exceptions.{ClientException, Neo4jException, ServiceUnavailableException, SessionExpiredException}
+import org.neo4j.driver.exceptions.{ClientException, Neo4jException, ServiceUnavailableException, SessionExpiredException, TransientException}
 import org.neo4j.driver.{Session, Transaction, Values}
 import org.neo4j.spark.service._
 import org.neo4j.spark.util.Neo4jUtil.closeSafety
@@ -78,7 +78,7 @@ abstract class BaseDataWriter(jobId: String,
     } catch {
       case neo4jTransientException: Neo4jException =>
         val code = neo4jTransientException.code()
-        if ((neo4jTransientException.isInstanceOf[SessionExpiredException] || neo4jTransientException.isInstanceOf[ServiceUnavailableException])
+        if (isRetryableException(neo4jTransientException)
           && !options.transactionMetadata.failOnTransactionCodes.contains(code)
           && retries.getCount > 0) {
           retries.countDown()
@@ -93,6 +93,10 @@ abstract class BaseDataWriter(jobId: String,
     }
     Unit
   }
+
+  private def isRetryableException(neo4jTransientException: Neo4jException) = (neo4jTransientException.isInstanceOf[SessionExpiredException]
+      || neo4jTransientException.isInstanceOf[TransientException]
+      || neo4jTransientException.isInstanceOf[ServiceUnavailableException])
 
   private def logAndThrowException(e: Exception): Unit = {
     if (e.isInstanceOf[ClientException]) {

--- a/common/src/test/scala/org/neo4j/spark/util/Neo4jImplicitsTest.scala
+++ b/common/src/test/scala/org/neo4j/spark/util/Neo4jImplicitsTest.scala
@@ -94,9 +94,11 @@ class Neo4jImplicitsTest {
 
   @Test
   def `struct should return true if contains fields`: Unit = {
-    val struct = StructType(Seq(StructField("is_hero", DataTypes.BooleanType), StructField("name", DataTypes.StringType)))
+    val struct = StructType(Seq(StructField("is_hero", DataTypes.BooleanType),
+      StructField("name", DataTypes.StringType),
+      StructField("fi``(╯°□°)╯︵ ┻━┻eld", DataTypes.StringType)))
 
-    assertEquals(0, struct.getMissingFields(Set("is_hero", "name")).size)
+    assertEquals(0, struct.getMissingFields(Set("is_hero", "name", "fi``(╯°□°)╯︵ ┻━┻eld")).size)
   }
 
   @Test
@@ -104,5 +106,20 @@ class Neo4jImplicitsTest {
     val struct = StructType(Seq(StructField("is_hero", DataTypes.BooleanType), StructField("name", DataTypes.StringType)))
 
     assertEquals(Set[String]("hero_name"), struct.getMissingFields(Set("is_hero", "hero_name")))
+  }
+
+  @Test
+  def `getMissingFields should handle maps`: Unit = {
+    val struct = StructType(Seq(
+      StructField("im", DataTypes.StringType),
+      StructField("im.a", DataTypes.createMapType(DataTypes.StringType, DataTypes.StringType)),
+      StructField("im.also.a", DataTypes.createMapType(DataTypes.StringType, DataTypes.StringType)),
+      StructField("im.not.a.map", DataTypes.StringType),
+      StructField("fi``(╯°□°)╯︵ ┻━┻eld", DataTypes.StringType)
+    ))
+
+    val result = struct.getMissingFields(Set("im.aMap", "`im.also.a`.field", "`im.a`.map", "`im.not.a.map`", "fi``(╯°□°)╯︵ ┻━┻eld"))
+
+    assertEquals(Set("im.aMap"), result)
   }
 }

--- a/spark-2.4/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
+++ b/spark-2.4/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
@@ -14,6 +14,7 @@ import org.neo4j.spark.util.Neo4jOptions
 
 import java.time.ZoneOffset
 import scala.collection.JavaConverters._
+import scala.collection.mutable
 import scala.util.Random
 
 abstract class Neo4jType(`type`: String)
@@ -1264,5 +1265,54 @@ class DataSourceWriterTSE extends SparkConnectorScalaBaseTSE {
           |MERGE (u:User {id: event.id, firstName: event.firstName, lastName: event.lastName, gender: event.gender, birthday: event.birthday, email: event.email, userProp: obj.objStringField1})
           |""".stripMargin)
       .save()
+  }
+
+  @Test
+  def `should use property of a map property with a weird naming as value of "node.keys"`(): Unit = {
+    val df = Seq(
+      (1, Map("name" -> "Bonham", "instrument" -> "Drums")),
+      (2, Map("name" -> "Mayer", "instrument" -> "Guitar")),
+      (3, Map("name" -> "Bon Jovi", "instrument" -> "Voice"))
+    ).toDF("id", "foo.bar")
+
+    df.write
+      .mode("Overwrite")
+      .format(classOf[DataSource].getName)
+      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+      .option("labels", "Person")
+      .option("node.keys", "`foo.bar`.name:name,`foo.bar`.instrument:instrument_name")
+      .save()
+
+    val resultDf = ss.read
+      .format(classOf[DataSource].getName)
+      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+      .option("labels", "Person")
+      .load()
+
+    val expected = Set(
+      Map("name" -> "Bonham",
+        "instrument_name" -> "Drums",
+        "<labels>" -> mutable.WrappedArray.make(Array("Person")),
+        "id" -> 1L
+      ),
+      Map("name" -> "Mayer",
+        "instrument_name" -> "Guitar",
+        "<labels>" -> mutable.WrappedArray.make(Array("Person")),
+        "id" -> 2L
+      ),
+      Map("name" -> "Bon Jovi",
+        "instrument_name" -> "Voice",
+        "<labels>" -> mutable.WrappedArray.make(Array("Person")),
+        "id" -> 3L
+      )
+    )
+    val actual = resultDf.collect().map(row => {
+      Map("<labels>" -> row.getAs[Array[String]]("<labels>"),
+        "name" -> row.getAs[String]("name"),
+        "instrument_name" -> row.getAs[String]("instrument_name"),
+        "id" -> row.getAs[Long]("id")
+      )
+    }).toSet
+    assertEquals(expected, actual)
   }
 }

